src/abr/pensieve.hh

```diff
@@ -0,0 +1,75 @@
+#ifndef PENSIEVE_HH
+#define PENSIEVE_HH
+
+#include "abr_algo.hh"
+#include <stdio.h>
+#include <thread>
+#include <cstdint>
+#include <cstdlib>
+#include "ipc_socket.hh"
+#include "json.hpp"
+#include <endian.h>
+
+#include <deque>
+
+using json = nlohmann::json;
+
+
+class Pensieve : public ABRAlgo
+{
+public:
+  Pensieve(const WebSocketClient & client,
+      const std::string & abr_name, const YAML::Node & abr_config);
+
+  void video_chunk_acked(const VideoFormat & format,
+                         const double ssim,
+                         const unsigned int size,
+                         const uint64_t trans_time) override;
+  VideoFormat select_video_format() override;
+
+private:
+  /* past chunks and max number of them */
+  /*
+  size_t max_num_past_chunks_ {MAX_NUM_PAST_CHUNKS};
+  std::deque<Chunk> past_chunks_ {};
+  */
+
+  /* all the time durations are measured in sec
+  size_t max_lookahead_horizon_ {MAX_LOOKAHEAD_HORIZON};
+  size_t lookahead_horizon_ {};
+  double chunk_length_ {};
+  size_t dis_buf_length_ {MAX_DIS_BUF_LENGTH};
+  double unit_buf_length_ {};
+  size_t num_formats_ {};
+  double rebuffer_length_coeff_ {REBUFFER_LENGTH_COEFF};
+  double ssim_diff_coeff_ {SSIM_DIFF_COEFF};
+  */
+
+  /* for the current buffer length and quality */
+  size_t curr_buffer_ {};
+  size_t curr_format_ {};
+
+  size_t next_format_ {};
+  static FileDescriptor open_ipc() {
```

Comment:
Move this function to a member function (without `static`), or simply put it in the constructor.